File: library/builtins/traits/name.rs
use crate::*;
use std::collections::HashMap;
use wipple::*;

#[derive(TypeInfo, Debug, Clone)]
pub struct Name {
    pub name: String,
    pub location: Option<SourceLocation>,
}

impl Primitive for Name {}

impl Name {
    pub fn new(name: &str) -> Self {
        Name::new_located(name, None)
    }

    pub fn new_located(name: &str, location: Option<SourceLocation>) -> Self {
        Name {
            name: String::from(name),
            location,
        }
    }
}

stored_closure!(pub struct ComputeFn(&Env, &Stack) -> Result<Value>);

#[derive(Debug, Clone)]
pub enum Variable {
    Local(Value),
    Constant(Value),
    Computed(ComputeFn),
}

impl Variable {
    pub fn get_value(&self, env: &Env, stack: &Stack) -> Result<Value> {
        match self {
            Variable::Local(value) | Variable::Constant(value) => Ok(value.clone()),
            Variable::Computed(compute) => compute(env, stack),
        }
    }
}

#[derive(TypeInfo, Debug, Clone, Default)]
pub struct Variables(pub HashMap<String, Variable>);

env_key!(pub variables: Variables {
    visibility: EnvKeyVisibility::Public(UseMergeFn::merging(|parent: &mut Variables, new, _, _| {
        for (name, variable) in &new.0 {
            parent.0.entry(name.clone()).or_insert_with(|| variable.clone());
        }

        Ok(())
    })),
});

stored_closure!(pub struct AssignmentFn(VariadicInput, VariadicInput, &Env, &Stack) -> Result<()>);

impl Default for AssignmentFn {
    fn default() -> Self {
        AssignmentFn::new(|left, right, env, stack| {
            let left = Value::from(left);
            let right = Value::from(right);

            let name = left.get_or::<Name>("Expected name", env, stack)?;
            let value = right.evaluate(env, stack)?;

            env.set_variable(stack, &name.name, value.into_owned())?;

            Ok(())
        })
    }
}

env_key!(pub assignment: AssignmentFn {
    visibility: EnvKeyVisibility::Private,
});

stored_closure!(pub struct ComputedAssignmentFn(VariadicInput, VariadicInput, &Env, &Stack) -> Result<()>);

impl Default for ComputedAssignmentFn {
    fn default() -> Self {
        ComputedAssignmentFn::new(|left, right, env, stack| {
            let left = Value::from(left);
            let right = Value::from(right);

            let name = left.get_or::<Name>("Expected name", env, stack)?;

            let captured_env = env.child();

            env.set_computed_variable(
                stack,
                &name.name,
                ComputeFn::new(move |_, stack| {
                    catch!(Return in right.evaluate(&captured_env, stack).map(Cow::into_owned))
                }),
            )?;

            Ok(())
        })
    }
}

env_key!(pub computed_assignment: ComputedAssignmentFn {
    visibility: EnvKeyVisibility::Private,
});

macro_rules! set_variable {
    ($name:expr, $variable:expr, $env:expr, $stack:expr) => {{
        if let Some(variable) = get_variable_recursive($name, $env) {
            if matches!(variable, Variable::Constant(_)) {
                return Ok(()); // assignments to constants are ignored
            }
        }

        if $env.variables().0.contains_key($name) {
            return Err(error(
                &format!(
                    "Cannot assign to the same variable ('{}') more than once per scope",
                    $name,
                ),
                $stack,
            ));
        }

        $env.update_variables(|variables| {
            variables.0.insert($name.to_string(), $variable);
        });

        Ok(())
    }};
}

#[ext(pub, name = EnvVariablesExt)]
impl Env {
    fn set_variable(&self, stack: &Stack, name: &str, value: Value) -> Result<()> {
        set_variable!(name, Variable::Local(value), self, stack)
    }

    fn set_constant_variable(&self, stack: &Stack, name: &str, value: Value) -> Result<()> {
        set_variable!(name, Variable::Constant(value), self, stack)
    }

    fn set_computed_variable(&self, stack: &Stack, name: &str, compute: ComputeFn) -> Result<()> {
        set_variable!(name, Variable::Computed(compute), self, stack)
    }
}

impl Name {
    pub fn resolve(&self, env: &Env, stack: &Stack) -> Result<Value> {
        let variable = self.resolve_variable(env, stack)?;
        variable.get_value(env, stack)
    }

    pub fn resolve_if_present(&self, env: &Env, stack: &Stack) -> Result<Option<Value>> {
        let mut stack = stack.clone();
        stack
            .diagnostics_mut()
            .add(|| format!("Resolving variable '{}'", self.name));

        match self.resolve_variable_if_present(env) {
            Some(variable) => variable.get_value(env, &stack).map(Some),
            None => Ok(None),
        }
    }

    pub fn resolve_variable<'a>(&self, env: &'a Env, stack: &Stack) -> Result<Variable> {
        let mut stack = stack.clone();
        stack
            .diagnostics_mut()
            .add(|| format!("Resolving variable '{}'", self.name));

        self.resolve_variable_if_present(env).ok_or_else(|| {
            error(
                &format!("'{}' does not refer to a variable", self.name),
                &stack,
            )
        })
    }

    pub fn resolve_variable_if_present(&self, env: &Env) -> Option<Variable> {
        get_variable_recursive(&self.name, env)
    }
}

fn get_variable_recursive(name: &str, env: &Env) -> Option<Variable> {
    env.variables().0.remove(name).or_else(|| {
        env.parent()
            .and_then(|parent| get_variable_recursive(name, parent))
    })
}

pub(crate) fn setup(env: &Env, stack: &Stack) -> Result<()> {
    // Name : trait
    env.set_variable(stack, "Name", Value::of(Trait::of::<Name>()))?;

    // Name == Evaluate
    env.add_relation_between(stack, |name: Name| {
        EvaluateFn::new(move |env, stack| name.resolve(env, stack))
    })?;

    // Name == Text
    env.add_relation_between(stack, |name: Name| Text::new(name.name))?;

    Ok(())
}

File: Library/Sources/Traits/Name.swift
import Wipple

public struct Name: RawRepresentable, Hashable, Primitive {
    public var rawValue: String

    public init(rawValue: String) {
        self.rawValue = rawValue
    }
}

public extension Name {
    static var hole = Name(rawValue: "_")

    var isHole: Bool {
        self == .hole
    }
}

public enum Variable {
    public typealias Compute = (Env, Stack) throws -> Value

    case value(Value)
    case computed(Compute)
}

public extension Variable {
    func getValue(_ env: Env, _ stack: Stack) throws -> Value {
        switch self {
        case .value(let value):
            return value
        case .computed(let compute):
            return try compute(env, stack)
        }
    }
}

public extension Env {
    private struct VariablesKey: EnvKey {
        typealias Value = [String: Variable]

        static let visibility = EnvKeyVisibility<Value>.public { old, new, _, _ in
            old.merge(new, uniquingKeysWith: { old, _ in old })
        }
    }

    var variables: [String: Variable] {
        get { self[VariablesKey.self] ?? [:] }
        set { self[VariablesKey.self] = newValue }
    }
}

public struct Assign: Primitive {
    public typealias RawValue = (Value, Env, Stack) throws -> (String, Value)

    public let rawValue: RawValue

    public init(_ rawValue: @escaping RawValue) {
        self.rawValue = rawValue
    }

    public func callAsFunction(_ value: Value, _ env: Env, _ stack: Stack) throws -> (String, Value) {
        try self.rawValue(value, env, stack)
    }
}

public typealias HandleAssignment = Operator.OperationOf<ManyArity, ManyArity, Void>

public extension Env {
    private struct HandleAssignmentKey: EnvKey {
        typealias Value = HandleAssignment

        static let visibility = EnvKeyVisibility<Value>.private
    }

    var handleAssignment: HandleAssignment {
        get {
            self[HandleAssignmentKey.self] ?? { left, right, env, stack in
                guard left.count == 1 else {
                    throw Exit.error("Expected single value on left side of assignment", stack)
                }

                let assign = try left[0].get(Assign.self, or: "Cannot assign to this value because it does not have the Assign trait", env, stack)

                let (name, value) = try assign(Value(right), env, stack)

                try env.setVariable(name, to: value, stack)
            }
        }
        set { self[HandleAssignmentKey.self] = newValue }
    }
}

public extension Env {
    private struct HandleComputedAssignmentKey: EnvKey {
        typealias Value = HandleAssignment

        static let visibility = EnvKeyVisibility<Value>.private
    }

    var handleComputedAssignment: HandleAssignment {
        get {
            self[HandleComputedAssignmentKey.self] ?? { left, right, env, stack in
                guard left.count == 1 else {
                    throw Exit.error("Expected single value on left side of assignment", stack)
                }

                let name = try left[0].get(Name.self, or: "Expected name", env, stack)

                let capturedEnv = env.child()

                try env.setComputedVariable(name.rawValue, stack) { _, stack in
                    try catchReturn {
                        try Value(right).evaluate(capturedEnv, stack)
                    }
                }
            }
        }
        set { self[HandleComputedAssignmentKey.self] = newValue }
    }
}

public extension Env {
    func setVariable(_ name: String, to value: Value, _ stack: Stack) throws {
        try self.setVariable(name, to: .value(value), stack)
    }

    func setComputedVariable(_ name: String, to compute: @escaping Variable.Compute, _ stack: Stack) throws {
        try self.setComputedVariable(name, stack, compute: compute)
    }

    func setComputedVariable(_ name: String, _ stack: Stack, compute: @escaping Variable.Compute) throws {
        try self.setVariable(name, to: .computed(compute), stack)
    }

    private func setVariable(_ name: String, to variable: Variable, _ stack: Stack) throws {
        if Name(rawValue: name).isHole {
            return // assignments to '_' are ignored
        }

        if self.variables.keys.contains(name) {
            throw Exit.error("Cannot assign to a variable more than once in the same scope", stack)
        }

        self.variables[name] = variable
    }
}

public extension Name {
    func resolve(_ env: Env, _ stack: Stack) throws -> Value {
        var stack = stack
        stack.diagnostics.add("Resolving variable '\(self.rawValue)'")

        return try self.resolveVariable(env, stack, addDiagnostics: false).getValue(env, stack)
    }

    func resolveIfPresent(_ env: Env, _ stack: Stack) throws -> Value? {
        var stack = stack
        stack.diagnostics.add("Resolving variable '\(self.rawValue)'")

        guard let variable = self.resolveVariableIfPresent(env) else {
            return nil
        }

        return try variable.getValue(env, stack)
    }

    func resolveVariable(
        _ env: Env,
        _ stack: Stack,
        addDiagnostics: Bool = true
    ) throws -> Variable {
        var stack = stack

        if addDiagnostics {
            stack.diagnostics.add("Resolving variable '\(self.rawValue)'")
        }

        if self.isHole {
            return Variable.value(.empty)
        }

        guard let variable = self.resolveVariableIfPresent(env) else {
            throw Exit.error("'\(self.rawValue)' does not refer to a variable", stack)
        }

        return variable
    }

    func resolveVariableIfPresent(_ env: Env) -> Variable? {
        if let variable = env.variables[self.rawValue] {
            return variable
        }

        if let parent = env.parent {
            return self.resolveVariableIfPresent(parent)
        }

        return nil
    }
}

internal func setupName(_ env: Env, _ stack: Stack) throws {
    // Name : trait
    try env.setVariable("Name", to: Value(Trait(Name.self)), stack)

    // Name == Evaluate
    try env.addRelation(from: Name.self, to: Evaluate.self, stack) { name in
        Evaluate(name.resolve)
    }

    // Name == Assign
    try env.addRelation(from: Name.self, to: Assign.self, stack) { name in
        Assign { value, env, stack in
            (name.rawValue, try value.evaluate(env, stack))
        }
    }

    // Name == Text
    try env.addRelation(from: Name.self, to: Text.self, stack) { name in
        Text(rawValue: name.rawValue)
    }
}
